rfcs/0137-nix-language-version.md

```diff
@@ -0,0 +1,678 @@
+   If no version is specified in a bare Nix expression, assume the most recent language version supported by the evaluator.
+
+   This is to ease use of the REPL and evaluating ad hoc expressions.
+
+   <details><summary>Alternatives</summary>
+
+   - Make it mandatory
+       - (-) This will be very inconvenient to use in the REPL
+
+   - Assume the evaluator's current version
+       - (-) When the evaluator advances in language version, evaluation may fail on existing code
+       - (-) Defeats the purpose of explicit versioning: Which evaluator to use for a given file is left unspecified
+           - (-) Following the latest evaluator version may inadvertently break the code for older evaluators
+       - (+) Don't have to look up the latest version of the Nix language when writing code
+       - (+) Does not clutter the file names for what is supposed to be the latest version of the code
+
+   </details>
+
+1. Language versions prior to 6 are not supported.
+
+   <details><summary>Arguments</summary>
+
+   - (+) Does not require additional development effort
+   - (+) Prior langauge versions are not fully supported by current code already, and the rest of this proposal argues to deprecate old versions in the future in order to keep the implementation manageable
+   - (-) Legacy code will not get support for managing compatibility
+
+   </details>
+
+1. The Nix language evaluator provides a command to output the most recent Nix language version.
+   This command provides options to list all Nix language versions supported by the evaluator.
+
+   <details><summary>Arguments</summary>
+
+   - (+) This is for convenience to determine which features are available
+
+   </details>
+
+1. `builtins.langVersion` returns the language version used for evaluating the given expression.
+
+   <details><summary>Arguments</summary>
+
+   - (+) Could make use of it for generating Nix expressions programmatically and annotating them with the correct version
+   - (+) `builtins.langVersion` is already part of the stable interface, this way we can make it pure
+   - (-) Requires maintaining more API surface without a clear use case
+
+   </details>
+
+   <details><summary>Alternatives</summary>
+
+   - Return the latest language version instead
+
+      - (-) Doesn't help to determine what is used for evaluating the current expression
+          - (+) Might provide opportunities for forward-compatible Nix code
+          - (-) Brittle and defeats the purpose of this RFC
+      - (-) Impure, the value depends on the environment
+
+   - Don't expose `builtins.langVersion` at all
+
+      - (+) No need to have this if the Nix files are versioned
+      - (+) `builtins.langVersion` is not documented and not widely used
+      - (-) Requires a Nix language version bump to implement this RFC
+
+   </details>
+
+1. Each time the language specification (currently as embodied by the Nix language evaluator) is changed, the language version must be incremented.
```

Comment:
This is very imprecise IMO, but I already disagree with its spirit. I think that a bump in language version is needed if and only if a change is not backwards compatible: If existing nix code runs and evaluates the same way as before with that language change, bumping the version is not necessary. (I can formalize this for syntax and semantic if you want to)

The implication is that the language version feature will *not* be a replacement for "min nix version" tracking, since an old evaluator might not be able to deal with newer Nix code. This is in line with what other languages do, for example Rust has both MSRV and edition.

Reply:
> If existing nix code runs and evaluates the same way as before with that language change, bumping the version is not necessary.

I'm thinking in terms of such examples:

1. Nix evaluator added `builtins.foo` in Nix 2.99, a backward compatible change requiring no language version increment
2. Expression author uses `builtins.foo ? workaround` to keep their code working with Nix < 2.99
3. Expression author decides to drop support for Nix < 2.99 and now uses `builtins.foo` directly
4. Expression users running Nix < 2.99 get random breakages

This would be alleviated by instead requiring a version bump even on pure additions, as users would get an error message in place saying that their dependency needs a newer evaluator to run.

> The implication is that the language version feature will not be a replacement for "min nix version" tracking, since an old evaluator might not be able to deal with newer Nix code. 

Not sure what that means. By "language version feature" do you mean the output of `builtins.langVersion`, as opposed to declaring that the evaluator must support at least a given language version with `version <number>`? If so, as in the other comment, I'm not sure if `langVersion` should even continue to exist.

For reference: https://github.com/rust-lang/api-guidelines/discussions/231

Reply:
> Nix evaluator added builtins.foo in Nix 2.99, a backward compatible change requiring no language version increment

According to my (admittedly strict) definition above, this is not a backward compatible change, since existing code `builtins.foo ? workaround` from your example evaluates to a different result with it. So additions of new `builtins` usually require a version bump, while addition of e.g. new language operators does not.

> By "language version feature" do you mean the output of builtins.langVersion, as opposed to declaring that the evaluator must support at least a given language version with version <number>?

No, I was mainly referring to the `version` statement.

Reply:
> additions of new `builtins` usually require a version bump, while addition of e.g. new language operators does not.

How then would you even discern that a new operator was added? The same example applies. If adding an operator to the evaluator does not bump the version, the expression author can just use it without changing the version declaration, and that will break the expression for users using of a prior version of the evaluator.

Reply:
Yes, that's why I said:

> The implication is that the language version feature will *not* be a replacement for "min nix version" tracking, since an old evaluator might not be able to deal with newer Nix code. This is in line with what other languages do, for example Rust has both MSRV and edition.

Reply:
I still don't know what that means though. Can you expand?

Reply:
Keeping with the Rust example, new syntax features may be added at any time to the language, as long as they don't break old code. Of course when using those new syntax features, older Rust tooling may choke on that. That's why projects declare a "minimum supported Rust version", MSRV.

When making changes to the language the are *not* backwards compatible in that they might change the evaluation semantics of existing code (or more likely, make existing code not compile anymore), then this is done through the "Rust edition" declaration.

MSRV and Rust edition are two distinct features. The motivation here is that most of the language changes are pure additions which only require updating the toolchain, while breaking changes are bundled up into an edition bump every n years.

Reply:
The obvious downside of this split is that it ties the ecosystem to the tooling implementation, or at least its versioning. If we don't want that (see https://github.com/NixOS/rfcs/pull/137#discussion_r1226426777), we could discuss some alternatives:

1. (I think what the RFC currently suggests) Always bump the Nix version on any kind of change related to the language. I think if we didn't have to tie `builtins` to the language version this could maybe work, but I remain sceptical
2. Use `major.minor` for the Nix version, with the guarantee that minor bumps only require toolchain updates. This would require the tooling to check the minor version, and bail out if it doesn't support it.

Reply:
> According to my (admittedly strict) definition above, this is not a backward compatible change, since existing code builtins.foo ? workaround from your example evaluates to a different result with it. So additions of new builtins usually require a version bump, ...

I think we should reconsider this. IIUC the current stability guarantee is that `builtins` can grow new members without being considered a breaking change. I don't see any reason to roll this back. Doing `builtins.foo ? fallback` doesn't make sense if `foo` doesn't exist in any version.

Bumping the version would lock users out of the new function until they migrate to the new syntax which seems like an unnecessary burden. (Unless we keep `builtins.langVersion` and syntax version separate, in which case it doesn't hurt to increment `builtins.langVersion` as much as we want as it is just informational)

Reply:
Agreed, and I was just writing about this too. I still think that builtins should generally be part of the language due to how they work, however I would make an explicit exception to the definition of a breaking change such that introducing new builtins does *not* count. This might be a bit trickier to formalize, but definitely not impossible.

Reply:
I think it is pretty easy to formalize:

1. The existence and behaviour of any documented member of `builtins` is considered stable.
2. The behaviour and existence or lack there of of any other members are not considered stable and **must not** be relied upon.

I said "documented" here but we may loosen that a bit. (And in practice a member that is known to be used even if not documented will likely be treated as if it was, we just don't guarantee that).

I also don't know if we want to add other guarantees such as `bulitins.foo ? fallback` not returning an error? But I think we don't. Touch something that isn't declared and you are on your own in terms of stability. When we add new documented members we can guarantee (explicitly or implicitly) that accessing that name on old versions will return "doesn't exist" (or document whatever behaviour it had previously, for example if we documented a previously "internal" function).

Reply:
I gather from this, in the proposal we should explore in more detail if it's worth having a major-minor versioning and whether versioning builtins separately makes sense.

The key point why that matters is that breaking language changes require existing code to be changed to use new evaluators.

Reply:
@kevincox @piegamesde added an argument in favor of semver, as handling pure additions separately would simplify evaluator implementation. Since it's the only alternative without real drawbacks, it will be promoted to be part of the proposal. Will also add details on how this interacts with builtins: I think new builtins should be handled in the minor version, then the mechanism to determine compatibility stays uniform and we don't need eval-time tricks at the call site.